src/derivative_operators/BC_operators.jl

```diff
@@ -0,0 +1,160 @@
+abstract type AbstractBC{T} end
+
+# robin, general, and in general neumann BCs are all affine opeartors, meaning that they take the form Qx = Qax + Qb. neumann0 is not however; is a specialization needed?
+abstract type AffineBC{T,V} <: AbstractBC{T} end
+
+struct DirichletBC{T} <: AbstractBC{T}
+    l::T
+    r::T
+end
+
+struct NeumannBC{T, V<:AbstractArray{T}} <: AffineBC{T,V}
+    a_l::V
+    b_l::T
+    a_r::V
+    b_r::T
+    function NeumannBC(l::T, r::T, dx::AbstractArray{T}, order::T = one(T)) where {T,V}
+        s = calculate_weights(1, zero(T), zero(T):order) #generate derivative coefficients about the boundary of required approximation order
+
+        a_l = -dx_l.*s[2:end]./s[1]
+        a_r = -dx_r.*s[end:-1:2]./s[1]
+
+        b_l = dx_l.*l./s[1]
+        b_r = dx_r.*l./s[1]
+
+        return new{T, typeof(a_l)}(a_l, b_l, a_r, b_r)
+    end
+end
+
+struct PeriodicBC{T} <: AbstractBC{T}
+
+end
+
+"""
+  Implements a robin boundary condition operator Q that acts on a vector to give an extended vector as a result
+  Referring to (https://github.com/JuliaDiffEq/DiffEqOperators.jl/files/3267835/ghost_node.pdf)
+
+  the variables in correspond to al*u(0) + bl*u'(0) = cl
+
+  Write vector b̄₁ as a vertical concatanation with b0 and the rest of the elements of b̄ ₁, denoted b̄`₁, the same with ū into u0 and ū`. b̄`₁ = b̄`_2 = fill(β/Δx, length(stencil)-1)
+  Pull out the product of u0 and b0 from the dot product. The stencil used to approximate u` is denoted s. b0 = α+(β/Δx)*s[1]
+  Rearrange terms to find a general formula for u0:= -b̄`₁̇⋅ū`/b0 + γ/b0, which is dependent on ū` the robin coefficients and Δx.
+  The non identity part of Qa is qa:= -b`₁/b0 = -β.*s[2:end]/(α+β*s[1]/Δx). The constant part is Qb = γ/(α+β*s[1]/Δx)
+  do the same at the other boundary (amounts to a flip of s[2:end], with the other set of boundary coeffs)
+"""
+
+# For  condition, the variables correspond to al*u(0) + bl*u'(0) = cl
+struct RobinBC{T, V<:AbstractVector{T}} <: AffineBC{T,V}
+    a_l::V
+    b_l::T
+    a_r::V
+    b_r::T
+    function RobinBC(l::AbstractArray{T}, r::AbstractArray{T}, dx::AbstractArray{T}, order::T = one(T)) where {T,V}
+        cl, al, bl = l
+        cr, ar, br = r
+        dx_l, dx_r = dx
+
+        s = calculate_weights(1, zero(T), zero(T):convert(T,order)) #generate derivative coefficients about the boundary of required approximation order
+
+        a_l = -bl.*s[2:end]./(al .+ bl*s[1]./dx_l)
+        a_r = -br.*s[end:-1:2]./(ar .+ br*s[1]./dx_r)
+
+        b_l = cl/(al+bl*s[1]/dx_l)
+        b_r = cr/(ar+br*s[1]/dx_r)
+
+        return new{T, typeof(a_l)}(a_l, b_l, a_r, b_r)
+    end
+end
+
+"""
+Implements a generalization of the Robin boundary condition, where α is a vector of coefficients.
+Represents a condition of the form α[1] + α[2]u[0] + α[3]u'[0] + α[4]u''[0]+... = 0
```

Comment:
wow, nice generalization!

Reply:
Thanks! I'm well proud of that, it was super satisfying to see all the terms condense down like that.